runtime-sdk/src/context.rs

```diff
@@ -129,31 +140,54 @@ impl<'a> DispatchContext<'a> {
     /// Executes a function with the transaction-specific context set.
     pub fn with_tx<F, R>(&mut self, tx: transaction::Transaction, f: F) -> R
     where
-        F: FnOnce(TxContext<'_, '_>, transaction::Call) -> R,
+        F: FnOnce(Context<'_, '_>, transaction::Call) -> R,
     {
         // Create a store wrapped by an overlay store so we can either rollback or commit.
         let store = storage::MKVSStore::new(self.io_ctx.clone(), &mut self.runtime_storage);
         let store = storage::OverlayStore::new(store);
 
-        let tx_ctx = TxContext {
+        let tx_ctx = Context {
             mode: self.mode,
             runtime_header: self.runtime_header,
             runtime_round_results: self.runtime_round_results,
             store,
-            tx_auth_info: tx.auth_info,
+            tx_auth_info: Some(tx.auth_info),
             tags: Tags::new(),
             // NOTE: Since a limit is enforced (which is a u32) this cast is always safe.
-            message_offset: self.messages.len() as u32,
             max_messages: self.max_messages.saturating_sub(self.messages.len() as u32),
             messages: Vec::new(),
             values: &mut self.values,
         };
         f(tx_ctx, tx.call)
     }
+
+    /// Executes a function with function specific sub-context set.
+    pub fn with_ctx<F, R>(&mut self, f: F) -> R
+    where
+        F: FnOnce(Context<'_, '_>) -> R,
+    {
+        // Create a store wrapped by an overlay store so we can either rollback or commit.
+        let store = storage::MKVSStore::new(self.io_ctx.clone(), &mut self.runtime_storage);
+        let store = storage::OverlayStore::new(store);
+
+        let ctx = Context {
+            mode: self.mode,
+            runtime_header: self.runtime_header,
+            runtime_round_results: self.runtime_round_results,
+            store,
+            tx_auth_info: None,
+            tags: Tags::new(),
+            // NOTE: Since a limit is enforced (which is a u32) this cast is always safe.
+            max_messages: self.max_messages.saturating_sub(self.messages.len() as u32),
+            messages: Vec::new(),
+            values: &mut self.values,
+        };
+        f(ctx)
+    }
 }
 
-/// Per-transaction dispatch context.
-pub struct TxContext<'a, 'b> {
+/// Per-transaction/method dispatch sub-context.
+pub struct Context<'a, 'b> {
```

Comment:
I'd save this name for `trait Context`

for instance
```rust
pub trait Context {
    fn runtime_header(&self) -> &Header;
    fn runtime_round_results(&self) -> &RoundResults;
    fn runtime_state(&mut self) -> impl Store;
    fn emit_message(&mut self, msg: Message) -> Result<u32, Error>;
    fn value<V>(&mut self, key: &'static str) -> &mut V;
    fn take_value<V>(&mut self, key: &'static str) -> Box<V>;
}
```
at the very least, it'd prevent `DispatchContext` and `TxContext` from diverging (`emit_messages` vs `emit_message` is already telling)

Reply:
Yeah i also considered doing this, had some trouble getting the `runtime_state` lifetimes matching and all between the two existing implementations. But i don't think i tried returning the store as a trait object. Might do another go at it.

Reply:
I agree that generally having this as a trait is nicer. We can then have all of the module methods defined for the `Context` trait and don't need the added `with_ctx` method on the `DynamicContext`, and can just use it directly from `{begin,end}_block` and message hooks.

Reply:
I can take a whack at the lifetimes if you get stuck again. No guarantees that I can do any more than you, but at least we'll both feel more confident that it's impossible :)

Reply:
`impl Store` doesn't work as `impl Trait` is not allowed in trait methods.
Returning a trait object also doesn't work, as the `storage::Store` cannot be made into an object due to generic methods.

Using an associated type does look promising, but cannot get lifetimes to work: started with something like this
```
pub trait Context 
where
    Self::S: storage::Store,
{
    type S;

   fn runtime_state(&mut self) -> Self::S;
}

impl<'a, 'b> Context for TxContext<'a, 'b> 
{
    type S = &'b mut storage::OverlayStore<storage::MKVSStore<&'b mut &'a mut dyn mkvs::MKVS>>;

    fn runtime_state(&mut self) -> Self::S {
        self._runtime_state()
    }
}
```

The problem is that this `'b` lifetime in line `type S = &'b mut storage....` would actually need to be the elided lifetime from the `fn runtime_state<'this>(&mut self) -> Self::S`. Not sure if there's a way to specify that restriction in the `type S = ` line.

The workaround i came up with would be turning things around a bit:
```
pub trait Context 
where
    Self::S: storage::Store,
{
    type S;

   fn runtime_state(&mut self) -> &mut Self::S;
}

impl<'a, 'b> Context for TxContext<'a, 'b> 
{
    type S = storage::OverlayStore<storage::MKVSStore<&'b mut &'a mut dyn mkvs::MKVS>>;

    fn runtime_state(&mut self) -> &mut Self::S {
        self._runtime_state()
    }
}
```

This works, but then we would need to change something in the `DispatchContext` as the method there currently doesn't return a mutable reference: 
```
pub fn runtime_state(&mut self) -> storage::MKVSStore<&mut dyn mkvs::MKVS>
```
But it's not clear to me how, i guess i need to look up a bit more the storage implementations, which i tried to mostly ignore :-)

This looks like the best direction to go, or is there anything obvious i missed?

Reply:
See last commit, it introduces the `Context` trait, makes things much nicer, however there's a todo left for implementing the `runtime_state` for the `DispatchContext`

Reply:
> But it's not clear to me how, i guess i need to look up a bit more the storage implementations, which i tried to mostly ignore :-)

This is because the `MKVSStore` is keeping a `&mut mkvs::MKVS` internally. It is basically a thin wrapper around a `mkvs::MKVS` that just allows you to avoid the IO context everywhere.

Reply:
You could create one `storage::MKVSStore` instance in the constructor (instead of storing `&mut mkvs::MKVS` in `runtime_storage`) and then return a mut reference to it. Maybe?

Reply:
i think that worked! thanks!